assimp_py.c

```diff
@@ -1,519 +1,1178 @@
+            mat->mProperties[i]->mSemantic == type &&
+            mat->mProperties[i]->mIndex == index)
+        {
+            prop = mat->mProperties[i];
+            break;
+        }
+    }
+
+    // If the property wasn't found in the list, return None.
+    if (!prop) {
+        // fprintf(stderr, "DEBUG get_material_property: Property not found for key='%s', type=%u, index=%u\n", key, type, index);
+        Py_RETURN_NONE;
     }
-    return items;
+
+    // --- Use Fixed-Size Buffers for Array Getters ---
+    float fval[16];       // Fixed buffer for floats
+    int ival[16];         // Fixed buffer for ints
+    struct aiString sval; // For strings
+    unsigned int arr_size; // Input: buffer capacity, Output: elements written
+
+
+    // 2. Process based on the found property's type (prop->mType)
+    switch (prop->mType) {
+        case aiPTI_String:
+            // Use aiGetMaterialString (it directly uses the underlying prop->mData)
+            ret = aiGetMaterialString(mat, key, type, index, &sval);
+            if (ret == aiReturn_SUCCESS) {
+                py_value = PyUnicode_FromString(sval.data); // Use .data
+                if (!py_value) { ret = aiReturn_FAILURE; } // Check PyUnicode creation
+            }
+            break; // End String case
+
+        case aiPTI_Float:
+        case aiPTI_Buffer: // Treat buffer as float
+        case aiPTI_Double: // aiGetMaterialFloatArray handles double conversion internally
+             arr_size = 16; // Input: Our buffer capacity
+             // Call the generic Float Array getter
+             ret = aiGetMaterialFloatArray(mat, key, type, index, fval, &arr_size);
+             // fprintf(stderr, "DEBUG get_material_property (Float Fixed): key='%s', ret=%d, arr_size_written=%u\n", key, ret, arr_size);
+
+             if (ret == aiReturn_SUCCESS && arr_size > 0) { // Check success AND elements retrieved
+                 if (arr_size == 1) {
+                     py_value = PyFloat_FromDouble((double)fval[0]);
+                 } else {
+                     py_value = PyList_New(arr_size);
+                     if (py_value) {
+                         for (unsigned int i = 0; i < arr_size; ++i) {
+                             PyObject* f = PyFloat_FromDouble((double)fval[i]);
+                             if (!f) { Py_DECREF(py_value); py_value = NULL; break; }
+                             PyList_SET_ITEM(py_value, i, f); // Steals reference
+                         }
+                         if (!py_value && !PyErr_Occurred()) {
+                             PyErr_SetString(PyExc_RuntimeError, "Failed to build float list items");
+                         }
+                     } // else PyList_New failed
+                 }
+                 // Check object creation success
+                 if (!py_value && !PyErr_Occurred()) { ret = aiReturn_FAILURE; }
+                 else if (!py_value && PyErr_Occurred()) { ret = aiReturn_FAILURE; }
```

Comment:
```suggestion
                 if (!py_value) { ret = aiReturn_FAILURE; }
```
The result of `PyErr_Occured()` does not matter here, since in both possible cases the outcome is the same.